index/types.go

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+// Tag represents a metric tag.
+type Tag struct {
+	Key   []byte
+	Value []byte
+}
+
+// ID represents a metric identifier.
+type ID []byte
+
+// Index represents a reverse index.
+type Index interface {
+	// Insert inserts the given ID with provided tags.
+	Insert(ID, []Tag) error
+
+	// Remove removes the given ID.
+	Remove(ID) error
+
+	// Fetch retrieves the list of ids matching the given criterion.
+	Fetch(FetchOptions) []ID
+}
```

Comment:
I wonder if it makes sense to have separate `Reader` and `Writer` interfaces? And then combine them in the `Index` interface:

```
type Index interface {
    Writer
    Reader
}
```

My thinking is that it may be beneficial to distinguish when we need to read from the index, write to the index, or both.